### ISSUE TITLE
Potential issue with custom transformer masks when using fast path and batch_first=True

### ISSUE THREAD

**Issue by eduphil** [pytorch/pytorch#115472]
### 🐛 Describe the bug

When initializing a Transformer like this
```
nn.Transformer(hidden_dim * 2, 4, batch_first=True)
```

and then calling it like this
```
attention_mask = nn.Transformer.generate_square_subsequent_mask(x.shape[0])
output = self.transformer(x, x, src_mask=attention_mask, tgt_mask=attention_mask, src_is_causal=True, tgt_is_causal=True)
```

**it will work for training but not for predictions.** 

The error message for predictions is 
```shape '[1, 1, 52, 52]' is invalid for input of size 4096```
when the input shape of x is [64, 52, 20], where 64 is the batch dimension (in my case I'd like to ignore following batches using the given mask, which is why the mask is [64,64]).

**If I change batch_first to False, both training and predictions work.** I assume this could be an issue with the transformers internal  "fast path", which is only active during predictions, but not when training.

### Versions

I use PyTorch 2.1.

**Comment by mikaylagawarecki** [comment]
According to the documentation 


<img width="781" alt="Screenshot 2023-12-15 at 2 27 47 PM" src="https://github.com/pytorch/pytorch/assets/35276741/aec271b4-99e8-4f89-8107-e5f101f370c8">

The shape of `src_mask` should be `(S, S)` and `tgt_mask` should be `(T, T)`

where `S` should be `src.shape(1)` and `T` should be `tgt.shape(1)` when `batch_first=True`

So it makes sense to me that creating both masks with 
```
attention_mask = nn.Transformer.generate_square_subsequent_mask(x.shape[0])
```

would error for `batch_first=True`. Will look further into why this works for training
